Add confirmation prompt before moving groups to the recycling bin, Refactor MessageBox to allow custom buttons

kneitinger opened the pull request:
## **UPDATE**:
The scope of this PR changed quite a bit since it was open.  Originally it was to add a confirmation before moving groups to the recycle bin, which turned into modifying [all questions with destructive actions](https://github.com/keepassxreboot/keepassxc/pull/2376#issuecomment-433696137) to have buttons that reflect the operation instead of yes/no.  To implement that, MessageBox was rewritten to act analogously to how it did when it was `QMessageBox::StandardButton`-based, but allows for custom buttons.

The manual test cases I used [can be seen in this comment](https://github.com/keepassxreboot/keepassxc/pull/2376#issuecomment-436531333)

Below is the original PR, other requests can be found in the comments.

## Description
Spawn a yes/no QMessage box when "Delete Group" is selected on a group
that is not already in the recycle bin (note: the prompt for deletion
from the recycle bin was already implemented).  This follows the same
pattern and language as entry deletion.

## Motivation and context
Fixes #2125


## How has this been tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Group "TestGroup" added
- Right clicked and selected "Delete Group"
  - Prompt `Do you really want to move the group "TestGroup" to the recycle bin?` is shown
  -  ![2018-10-08_18-31-34](https://user-images.githubusercontent.com/5200550/46643544-66015700-cb31-11e8-8f19-c2f02db72d7e.png)
  - Chose "No", group was not moved to recycle bin.
  - Repeated process and chose "Yes", group was moved to recycle bin
  - Repeated above steps from top of window menu bar -> Groups -> Delete Group, same results
- Ran existing tests

## Types of changes
<!--- What types of changes does your code introduce? -->
<!--- Please remove all lines which don't apply. -->
- ✅ Bug fix (non-breaking change which fixes an issue)

## Checklist:
<!--- Please go over all the following points. -->
<!--- Again, remove any lines which don't apply. -->
<!--- Pull Requests that don't fulfill all [REQUIRED] requisites are likely -->
<!--- to be sent back to you for correction or will be rejected.  -->
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

phoerious commented:
TBH, I am not a big fan of Yes/No boxes. There is almost never a situation where they are good UX design. I vote for changing the buttons to"Delete" and "Cancel".

kneitinger commented:
@phoerious I could be into that. I made this consistent with the existing dialogs for moving an entry to the recycling bin, so I would want to change those as well.

droidmonkey commented:
I do agree with @phoerious on this one. It makes a lot of sense, especially for quick questions, to have the (complete) answer as the choice in the button. Under Material design rules, you would also make the delete button red, but that is overkill. I think most of our dialogs will go this route, perhaps we should extend the MessageBox class to include these additional choices. See: http://doc.qt.io/qt-5/qmessagebox.html#advanced-usage

kneitinger commented:
@phoerious @droidmonkey I went through the `src` dir and found all of the occurrences of prompts in which choices are present. Some of them match the action description style that we're moving towards, but the majority of them are Yes/No.  Happy to go through and fix all of these for this PR, unless you feel that it would make more sense to put them into their own PR:

--- 

### Occurrences of `QMessageBox` questions

`gui/DatabaseWidget.cpp`:

- 623: execute URL if command:  Y/N
- 670: deleting group: Y/N
- 1248: db file has changed, reload?: Y/N
- 1271: db file change and you have unsaved changes, merge?: Y/N
- 1489: Empty recylce bin?: Y/N

`gui/EditWidgetIcons.cpp`:

- 441: confirm deletion of icon being used by other entries: Y/N

`gui/DatabaseTabWidget.cpp`:

- 280: is in edit mode, discard changes and close anyway?: Discard/Cancel
- 297: DB was modified, save changes? Yes/Discard/Cancel
- 387: Disable safe saves and try to save again?: Y/N
- 737: Can lock db, since currently editing: Discard(which locks db)/Cancel
- 750: Locking but DB modified. save before lock?: Save/Discard/Cancel

`browser/BrowserService.cpp`:

- 203: shared encryption key already exists, overwrite?: Y/N
- 362: update entry information?: Y/N

`gui/EditWidgetProperties.cpp`:

- 69: Delete plugin data?: Y/Cancel

`autotype/AutoType.cpp`:

- 739: long delay; proceed?:  Y/N
- 748: slow keypresses; proceed?:  Y/N
- 759: args repeated very often; proceed?:  Y/N

`gui/entry/EditEntryWidget.cpp`:

- 809: apply generated pw to entry?: Y/N
- 932: entry has unsaved changes: Cancel/Save/Discard
- 1039: confirm removal of attribute: Y/N

`gui/entry/EntryAttachmentsWidget.cpp`:

- 166: confirm removal of attachment: Y/N
- 213: confirm overwrite of existing file with attachment: Y/N/Cancel

`gui/dbsettings/DatabaseSettingsWidgetMasterKey.cpp`:

- 178: continue without a password warning: Y/Cancel

`gui/dbsettings/DatabaseSettingsWidgetEncryption.cpp`:

- 250: very high number of rounds warning argon2: "Understood, keep number"/Cancel
- 267: very high number of rounds warning aes-kdf: "Understood, keep number"/Cancel

kneitinger commented:
And to touch on @droidmonkey's last comment:

I'm not sure I have the eye to make delete-centric buttons red without making them look like trash :slightly_smiling_face: 

It seems there are some patterns visible in occurrences listed in my last comment. I agree it would be nice to add some `MessageBox` functions to cover some of the common cases.

droidmonkey commented:
Which ones do you recommend changing? Most of them seem fine to me (like your comment suggests).

droidmonkey commented:
Great suggestions, I agree with them. Wherever possible, prefer the use of a single verb. Sorry for the confusion in earlier comments.

droidmonkey commented:
@kneitinger Do you want to introduce the new buttons in a different PR or add to this one?

kneitinger commented:
@phoerious @droidmonkey I went through and modified all of the prompts

## Verification
- `browser/BrowserService.cpp`:
  - 203: **shared encryption key already exists, overwrite?:** ~~Y/N~~ _Overwrite/Cancel_
    - Do not know how to test
  - 362: **update entry information?:** ~~Y/N~~ _Save/Cancel_
    - Do not know how to test
- `gui/DatabaseTabWidget.cpp`:
  - 297: **DB was modified, save changes?** ~~Yes/Discard/Cancel~~ Save/Discard/Cancel
    - Disabled autosave, created a new entry
    - Attempted to close application
    - Correct prompt was displayed 
![2018-11-06_22-38-27](https://user-images.githubusercontent.com/5200550/48116674-af6dc080-e21b-11e8-86ae-1efa18ffd2e5.png)

  - 387: **Disable safe saves and try to save again?:** ~~Y/N~~ _Disable Safe Saves/Cancel_
    - Copied test database to a new directory and opened it
    - Deleted directory that contains new db
    - Attempted to save db 3 times
    - Correct prompt displayed, and upon selecting "Disable", the safe save setting was disabled 
![2018-11-06_22-40-57](https://user-images.githubusercontent.com/5200550/48116693-bf85a000-e21b-11e8-832a-27a29032f1ec.png)

- `gui/DatabaseWidget.cpp`:
  - 472: **delete entry permanently:** ~~Y/N~~ _Delete/Cancel_ ![2018-11-06_22-48-03](https://user-images.githubusercontent.com/5200550/48116812-168b7500-e21c-11e8-9541-fb2ca74ddf70.png)
  - 490: **move entry to recycle bin:** ~~Y/N~~ _Move to Recycle Bin/Cancel_ 
![2018-11-06_22-47-54](https://user-images.githubusercontent.com/5200550/48116806-0e333a00-e21c-11e8-94c9-af3b4600742c.png)
  - 682: **delete group permanently:** ~~Y/N~~ _Delete/Cancel_ 
![2018-11-06_22-47-24](https://user-images.githubusercontent.com/5200550/48116782-04a9d200-e21c-11e8-8212-493aa085e32b.png)
  - 691: **move group to recycle bin:** ~~Y/N~~ _Move to Recycle Bin/Cancel_ 
![2018-11-06_22-48-14](https://user-images.githubusercontent.com/5200550/48116858-3458da00-e21c-11e8-94e7-383f98b5b8f9.png)
  - 1271: **db file change and you have unsaved changes, merge?:** ~~Y/N~~ _Merge/Cancel_
    - `cp testdb.kdbx newdb.kdbx`
    - Opened testdb.kdbx with autosave disabled
    - Created new entry
    - `mv newdb.kdbx testdb.kdbx`
    - Correct prompt displayed, and upon selecting "Merge", console log shows successful merge 
![2018-11-06_22-45-46](https://user-images.githubusercontent.com/5200550/48116742-e512a980-e21b-11e8-9c64-05071f82d25b.png)

  - 1489: **Empty recycle bin?:** ~~Y/N~~ _Empty Recycle Bin/Cancel_ 
![2018-11-06_22-48-26](https://user-images.githubusercontent.com/5200550/48116870-420e5f80-e21c-11e8-9ce5-138f202b8b9e.png)

- `gui/EditWidgetIcons.cpp`:
  - 441: **confirm deletion of icon being used by other entries:** ~~Y/N~~ _Delete/Cancel_
    - Create new entry
    - Go to Icon edit tab and select a preexisting custom icon
    - Click "Delete custom icon"
    - Correct prompt displayed, and upon selecting "Delete", custom icon is removed 
![2018-11-06_22-55-44](https://user-images.githubusercontent.com/5200550/48116885-4cc8f480-e21c-11e8-90ce-1b11292f4004.png)

- `gui/EditWidgetProperties.cpp`:
  - 69: **Delete plugin data?:** ~~Y/Cancel~~ _Delete/Cancel_
    - Do not know how to populate plugin data or test
- `gui/entry/EditEntryWidget.cpp`:
  - 1039: **confirm removal of attribute:** ~~Y/N~~ _Remove/Cancel_
    - Add new attribute
    - Select new attribute and click "Remove"
    - Correct prompt displayed, and upon selecting "Remove", the attribute is removed 
![2018-11-06_22-57-38](https://user-images.githubusercontent.com/5200550/48116900-56eaf300-e21c-11e8-8ed4-1a4a624d95fa.png)

- `gui/entry/EntryAttachmentsWidget.cpp`:
  - 166: **confirm removal of attachment:** ~~Y/N~~ _Remove/Cancel_
    - Edit entry and add two attachments
    - Select one entry and click "Remove"
    - Correct prompt displayed, and upon selecting "Remove", the attachment is removed
    - Repeat above process with multiple attachments and verify correct number and removal 
![2018-11-06_22-59-42](https://user-images.githubusercontent.com/5200550/48116922-610cf180-e21c-11e8-909b-b6382dfebed7.png)

  - 213: **confirm overwrite of existing file with attachment:** ~~Y/N/Cancel~~ _Overwrite/Skip/Cancel_
    - Edit entry and add two attachments from home directory
    - Select one entry and click "Save"
    - Choose home directory in file select dialog
    - Correct prompt displayed, showing only "Overwrite" and "Cancel", upon selecting overwrite, attachment is saved. 
![2018-11-06_23-04-57](https://user-images.githubusercontent.com/5200550/48116959-76821b80-e21c-11e8-9efa-8d20ab29720f.png)
    - Select both entries and click "Save"
    - Correct prompt displayed, showin "Overwrite", "Skip", and "Cancel", upon selecting overwrite, attachment is saved, and next file's prompt is shown. Upon selecting skip, attachment is not saved, and next file's prompt is shown. Upon selecting cancel, no more overwrite dialogs are shown, regardless of index of current attachment. 
![2018-11-06_23-04-48](https://user-images.githubusercontent.com/5200550/48116966-7da92980-e21c-11e8-993f-e20f18a7bf89.png)

kneitinger commented:
Apologies, forgot to update the GUI tests after these changes, I'll take care of that.  As for the CodeFactor test, it deemed a function whose only edit I made was removing a single trailing space to be "Complex".  I wouldn't mind refactoring it, but do you know of a way to approximate CodeFactor's behavior locally so I don't need to push to see if it's sufficient?

droidmonkey commented:
Don't worry about codefactor

kneitinger commented:
@droidmonkey Thanks, I noticed the same thing.  I opted to do it out of MessageBox due to difficulty coming up with nice generic solutions. I didn't see an issue because nextAnswer didn't seem to be used in `src`, but now see how its used in the tests.  I think I have a couple ideas on how to make this work neatly, but will definitely need some feedback. I'll post what I come up with in a comment.

kneitinger commented:
@droidmonkey 

I came up with a rough sketch of something that seems to work well: [branch: kneitinger/messagebox_revamp](https://github.com/kneitinger/keepassxc/commit/1f8da88ce6c5cbeb6eba3ed7f8f2831d20a20b17#diff-d04e4784799ae513658f4eda2faa1b0cR459-d04e4784799ae513658f4eda2faa1b0cR460
)

The main focus was making the interactions that happen outside of the `MessageBox` files (such as in `src/guiDatabaseWidget.cpp`) just as nice as they were before this PR.  Essentially, if we went this route, every `QMessageBox::{Ok,Cancel,Save,...}` would just need to be changed to `MessageBox::{Ok,Cancel,Save,...}`.  At the same time, new buttons can be added application-wide with just a few lines of code.

Breakdown of the changes:
- `src/gui/MessageBox.h`:
  - 28-59: Defines enum that works analogously to QMessageBox::StandardButtons
  - 96: Map to store info of button pointers to MessageBox::Button so that return type
      of MessageBox::newQuestion can just be MessageBox::Button instead of a pointer. This is needed so that developers can use conditionals such as `if (answer == MessageBox::Overwrite)` with buttons other than `QMessageBox::StandardButton`s
- `src/gui/MessageBox.cpp`:
  - 128-198: Create/Add a button that represents the intent of the `MessageBox::Button` value.
             Since `QMessageBox::clickedButton()` returns a pointer to a button, we'll use a map
             to track which button corresponds to which `MessageBox::Button`
  - 75-79: Start building up a `QMessageBox` to act as a question
  - 81-85: iterate over possible buttons and when a match is found, add it to the `QMessageBox`
  - 88: show `QMessageBox` and wait for user input
  - 90-92: Look up the returned pointer to get its corresponding `MessageBox::Button` and return it.
  - 95-98: Handle case when m_newNextAnswer is set (for guitest)
- `src/gui/DatabaseWidget.cpp`:
  - 459-462: shows standard usage.  Works just like [how MessageBox::Question worked before](https://github.com/kneitinger/keepassxc/blob/d6ffee5e99905a3bf237fc554b5e9271097cc849/src/gui/DatabaseWidget.cpp#L468-L469),
    but with MessageBox::Button buttons instead of QMessageBox::StandardButtons.  e.g.
    MessageBox::Ok vs. QMessageBox::Ok
  - 468-473: quick way to show that the new model has setNextAnswer functionality.
    When ran, application does not exit with code 22 meaning the answer was successfully set.

I'd like to hear your thoughts on this.  Keep in mind that this is just a rough prototype for illustrating an approach, and a final solution using this approach would be a lot cleaner (possibly using Q_FLAGS on enum, Refactoring for loops out of `question(), warning(), ...` and into  `setButtons()`, etc.)

Thanks in advance!

droidmonkey commented:
Nice! That is exactly what I had in mind. For the button map, is that translation ready? You may need to add the "tr()" wrappers in there.

droidmonkey commented:
I would just make MessageBox a QObject to simplify.

kneitinger commented:
@droidmonkey I worked in the translation on the new buttons as well as using the already translated text from the QT standard buttons. I accrued a CodeFactor issue for the `MessageBox::addButton()` function, which I'd like to address; as well as some formatting I'd like to improve upon seeing the diff in GitHub. I'll let you know when thats all pushed up :+1:

droidmonkey commented on src/main.cpp:
```diff
@@ -103,6 +103,9 @@ int main(int argc, char** argv)
     }
 
+    // Populate MessageBox button labels with text from QMessageBox
+    MessageBox::initializeButtonDefs();
```
This should occur in the Bootstrap::bootstrapApplication() function above

kneitinger commented:
I've moved the `initializeButtonDefs()` call, and corrected some of the dialogs that I overlooked when merging in the Database Widget refactor commits.

I went through and re-ran my manual tests listed in previous comments, and its working the same, however, in reimplementing the buttons, I've overlooked the icons:
- Before:
![48116674-af6dc080-e21b-11e8-86ae-1efa18ffd2e5](https://user-images.githubusercontent.com/5200550/49610238-dde1d700-f952-11e8-8838-d092fcd24bf3.png)
- Current:
![2018-12-06_12-30-32_iconlesss](https://user-images.githubusercontent.com/5200550/49610249-e9350280-f952-11e8-8061-bca756a895a5.png)

If we want them back, It would not be much work to obtain the icons for `StandardButton`s in the same way I retrieve the label text for them.  Let me know what you think.

droidmonkey commented:
Oh crap, its because Windows.h is included. You have to add a ```#undef MessageBox``` towards the top.

As for the icons, I never noticed them before! They do not appear on Windows.

![image](https://user-images.githubusercontent.com/2809491/49612052-81da7b00-f971-11e8-9c83-90a67a43a70c.png)

kneitinger commented:
:laughing: Yeah, I thought the lack of icons was a classier experience, but wasn't sure if they were kept for some reason. I haven't changed my QT theme because I though the icons would stay and look out of place in a better looking theme.  After seeing your screenshot, tried changing the theme and wow turns out KeePassXC can look so good with a modern theme and no stock icons!

Implemented the `#undef` and verified it builds and functions correctly on a Windows env I finally set up.  Unless anyone has any changes they would like to see, I believe the work on this PR is done now!

kneitinger commented:
Updated the PR title and added "Update" section to the top of the PR description to provide some context for any new viewers.

kneitinger commented:
Hi all, just a bump that this is ready for further review, or merge if ready :slightly_smiling_face:

kneitinger commented:
Thanks, congrats on completing your finals.

droidmonkey commented:
Thanks! Fantastic work on this PR.
